contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,555 @@
+    }
+
+    /**
+     * ONLY OPERATOR: Updates TransformInfo entry for a transform component. Must already have been set before updating. If not,
+     * use setTransformInfo.
+     *
+     * @param _transformComponent       component that needs to be transformed/untransformed during rebalance
+     * @param _transformInfo            new TransfomInfo entry for the transform component
+     */
+    function updateTransformInfo(address _transformComponent, TransformInfo memory _transformInfo) external onlyOperator {
+        require(
+            transformComponentInfo[_transformComponent].underlyingComponent != address(0),
+            "TransformInfo not set yet"
+        );
+
+        // Maximum transform size is limited to MAX_UINT_96 to prevent overflows when multiplying this maximum size by
+        // the transform components exchange rate to get a maximum untransform size. This prevents using MAX_UINT_256 
+        // to represent an unlimited amount.
+        require(_transformInfo.maxTransformSize <= type(uint96).max, "max transform size must be less than MAX_UINT_96");
+        transformComponentInfo[_transformComponent] = _transformInfo;
+    }
+
+    /**
+     * ONLY OPERATOR: Begins a rebalance. Must supply the set components as well as the target unit allocation for each. If the
+     * component is a transform component, then the target units should be measured in the equivalent value of the underlying tokens
+     *
+     * @param _setComponents            array of components involved in rebalance including components being removed (target units set to 0)
+     * @param _targetUnitsUnderlying    array of target units at end of rebalance, maps to same index of _components array
+     */
+    function startIPRebalance(address[] memory _setComponents, uint256[] memory _targetUnitsUnderlying) external onlyOperator {
+        require(_setComponents.length == _targetUnitsUnderlying.length, "length mismatch");
+        stage = RebalanceStage.UNTRANSFORM;
+
+        // Save rebalanceParams
+        for (uint256 i = 0; i < _setComponents.length; i++) {
+            targetUnitsUnderlying[_setComponents[i]] = _targetUnitsUnderlying[i];
+        }
+
+        setComponentList = _setComponents;
+    }
+
+    /**
+     * ONLY ALLOWED CALLER: Untransforms components. This function must be called after starting the rebalance, but before calling startTrades.
+     * The untransformData parameter can be fetched by the rebalance bots by calling getUntransformData on the relevent TransformHelper.
+     *
+     * @param _transformComponents      array of components to untransform
+     * @param _untransformData          array of untransform data 
+     */
+    function batchUntransform(
+        address[] memory _transformComponents,
+        bytes[] memory _untransformData
+    )
+        external
+        onlyAllowedCaller(msg.sender)
+    {
+        require(_transformComponents.length == _untransformData.length, "length mismatch");
```

Comment:
Would suggest adding a "NONE" element in ENUM and checking it explicitly here, to verify we are in the right state.

Reply:
In general, would suggest doing the following in every function,
1. Verify we are in the right initial state, using the RebalanceStage function
2. Check other requirements and perform the actions corresponding to the function
3. Update the state to the next possible state

Reply:
There is unfortunately some problems with verifying that a stage has been completed on-chain. Since some transform/untransform steps can have slippage, it may be impossible to hit the target units to transform and untransform. This makes it difficult to check if the stage has been completed. We could add a percentage factor, so we could instead mark a stage as complete if we are within x% of the target. However, this would come at a significant gas cost, since we would now need to fetch all target units and position units for every single transform/untransform call. This will add a significant number of `SLOAD` and `CALL` operations, which tend to be pricey. We would run into the same problem when trying to check if the GeneralIndexModule rebalance is complete.

Reply:
Also there is no `trade` function in this extension since we just use the `GeneralIndexModule` to do the trades. We could add a trade function to this extension that just checks the stage and forwards the call to `GeneralIndexModule`, but it still would not prevent an allowed caller for GIM from bypassing the check.